### PR TITLE
EN-6151 - integrate elrond in blockatlas

### PR DESCRIPTION
**Pull request by miiu96**
Elrond Proxy URL : `https://api.elrond.com`

In order to integrate Elrond in blockatlas we have used the following API calls against our proxy:
 - ` /network/status/4294967295`: gets the latest block nonce (number) - for method `CurrentBlockNumber()`
 -  `/block/4294967295/:nonce`: gets a block by nonce (number) - for method `GetBlockByNumber()`
 -  `/address/:accountAddress/transactions`: gets latest transactions of a given account - for method `GetTxsOfAddress()`

### CODE REVIEW

File: config.yml [hunk 1]
```diff
@@ -219,3 +219,7 @@ solana:
 
 near:
   api: https://staging-rpc.nearprotocol.com
+
+# public proxy : https://wallet-api.elrond.com
+elrond:
+  api: http://localhost:8001
```
Comment (AdoAdoAdo) [inline config.yml]: update the proxy?

File: config.yml [hunk 2]
```diff
@@ -219,3 +219,7 @@ solana:
 
 near:
   api: https://staging-rpc.nearprotocol.com
+
+# public proxy : https://wallet-api.elrond.com
```
Comment (andreibancioiu) [inline config.yml]:
```
https://api.elrond.com
```

We will use this one as the official API endpoint.

File: platform/elrond/model.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	GasPrice      uint64        `json:"gasPrice"`
+	GasLimit      uint64        `json:"gasLimit"`
+	Data          string        `json:"data"`
+	Signature     string        `json:"signature"`
+	Timestamp     time.Duration `json:"timestamp"`
+	Status        string        `json:"status"`
+}
+
+func (t *Transaction) Fee() string {
+	fee := big.NewInt(0).SetUint64(t.GasPrice)
```
Comment (andreibancioiu) [inline platform/elrond/model.go]: The formula doesn't always hold. Can we get it from the proxy instead? That is, don't compute it here?

File: platform/elrond/model.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	fee.Mul(fee, big.NewInt(0).SetUint64(t.GasLimit))
+
+	return fee.String()
+}
+
+func (t *Transaction) Stratus() blockatlas.Status {
```
Comment (andreibancioiu) [inline platform/elrond/model.go]: Typo, status.

File: platform/elrond/transaction.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		Meta: blockatlas.Transfer{
+			Value:    blockatlas.Amount(srcTx.Value),
+			Symbol:   coin.Elrond().Symbol,
+			Decimals: coin.Elrond().Decimals,
+		},
+	}
+	if address != "" {
+		tx.Direction = srcTx.Direction(address)
+	}
+	if srcTx.Sender == metachainID {
```
Comment (andreibancioiu) [inline platform/elrond/transaction.go:49]: Just add a comment to state and explain that logic of ours.

File: config.yml (hunk 1 above)
Comment (AdoAdoAdo) [reply config.yml]: also update the readme with Elrond as supported coin (png)

File: platform/elrond/model.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	Round         uint64        `json:"round"`
+	Value         string        `json:"value"`
+	Receiver      string        `json:"receiver"`
+	Sender        string        `json:"sender"`
+	ReceiverShard uint32        `json:"receiverShard"`
+	SenderShard   uint32        `json:"senderShard"`
+	GasPrice      uint64        `json:"gasPrice"`
+	GasLimit      uint64        `json:"gasLimit"`
+	Data          string        `json:"data"`
+	Signature     string        `json:"signature"`
```
Comment (AdoAdoAdo) [inline platform/elrond/model.go]: is there a need to include the signature? I don't see much point to it being included as there is no sig verification done and and not used as identification anywhere, so I think it can be removed.

File: platform/elrond/transaction_test.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	"round":35462,
+	"value":"82516976060558456822",
+	"receiver":"erd10yagg2vme2jns9zqf9xn8kl86fkc6dr063vnuj0mz2kk2jw0qwuqmfmaw0",
+	"sender":"4294967295",
+	"receiverShard":0,
+	"senderShard":4294967295,
+	"gasPrice":0,
+	"gasLimit":0,
+	"data":"",
+	"signature":"",
```
Comment (AdoAdoAdo) [inline platform/elrond/transaction_test.go:21]: if we remove the signature from tx model, remove it from here as well

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: platform/registry.go [hunk 8]
```diff
@@ -2,6 +2,7 @@ package platform
 
 import (
 	"fmt"
+
```
Comment (andreibancioiu) [inline platform/registry.go]: If possible undo this change, but optional :)

File: platform/elrond/block.go [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package elrond
+
+import "github.com/trustwallet/blockatlas/pkg/blockatlas"
+
+func (p *Platform) CurrentBlockNumber() (int64, error) {
+	return p.client.GetCurrentBlock()
```
Comment (andreibancioiu) [inline platform/elrond/block.go]: Rename to `p.client.CurrentBlockNumber()`? (like Cosmos does).

File: platform/elrond/client.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	"fmt"
+
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+type Client struct {
+	blockatlas.Request
+}
+
+func (c *Client) GetCurrentBlock() (num int64, err error) {
```
Comment (andreibancioiu) [inline platform/elrond/client.go]: Rename like Cosmos?

File: platform/elrond/client.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	var latestNonce LatestNonce
+	err = c.Get(&latestNonce, "block/latest-nonce", nil)
+	if err != nil {
+		return num, err
```
Comment (andreibancioiu) [inline platform/elrond/client.go]: Return `0, err` (like Cosmos)?

File: platform/elrond/client.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	path := fmt.Sprintf("block/meta/%d", uint64(height))
+	err := c.Get(&block, path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &blockatlas.Block{
+		Number: int64(block.Nonce),
+		ID:     block.Hash,
+		Txs:    NormalizeTxs(block.Transactions, ""),
```
Comment (andreibancioiu) [inline platform/elrond/client.go]: `txs := NormalizeTxs(block.Transactions, "")` in advance (a bit more readable).

File: platform/elrond/client.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+func (c *Client) GetTxsOfAddress(address string) (blockatlas.TxPage, error) {
+	var bulkTxs BulkTransactions
+	path := fmt.Sprintf("address/%s/transactions", address)
+	err := c.Get(&bulkTxs, path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return NormalizeTxs(bulkTxs.Transactions, address), nil
```
Comment (andreibancioiu) [inline platform/elrond/client.go]: Separate variable `txs` in advance?

File: platform/elrond/model.go [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+	Nonce uint64 `json:"nonce"`
+}
+
+type Block struct {
+	Nonce        uint64        `json:"nonce"`
+	Hash         string        `json:"hash"`
+	Transactions []Transaction `json:"transactions"`
+}
+
+type BulkTransactions struct {
```
Comment (andreibancioiu) [inline platform/elrond/model.go]: Perhaps rename to `TransactionsPage` (like Cosmos)?

File: platform/elrond/model.go (hunk 4 above)
Comment (andreibancioiu) [inline platform/elrond/model.go]: `tx` not `t`.

File: platform/elrond/client.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	return &blockatlas.Block{
+		Number: int64(block.Nonce),
+		ID:     block.Hash,
+		Txs:    NormalizeTxs(block.Transactions, ""),
+	}, nil
+}
+
+func (c *Client) GetTxsOfAddress(address string) (blockatlas.TxPage, error) {
+	var bulkTxs BulkTransactions
+	path := fmt.Sprintf("address/%s/transactions", address)
```
Comment (andreibancioiu) [inline platform/elrond/client.go:49]: Just add a comment that we currently fetch a number of `...?...` transactions. But that we may add pagination in the future.

File: platform/elrond/transaction.go [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+const metachainID = "4294967295"
+
+func (p *Platform) GetTxsByAddress(address string) (blockatlas.TxPage, error) {
+	return p.client.GetTxsOfAddress(address)
+}
+
+func NormalizeTxs(srcTxs []Transaction, address string) (txs []blockatlas.Tx) {
```
Comment (andreibancioiu) [inline platform/elrond/transaction.go:15]: Add comment?

File: platform/elrond/transaction.go [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+		From:     srcTx.Sender,
+		To:       srcTx.Receiver,
+		Fee:      blockatlas.Amount(srcTx.Fee()),
+		Status:   srcTx.Stratus(),
+		Sequence: srcTx.Nonce,
+		Meta: blockatlas.Transfer{
+			Value:    blockatlas.Amount(srcTx.Value),
+			Symbol:   coin.Elrond().Symbol,
+			Decimals: coin.Elrond().Decimals,
+		},
```
Comment (andreibancioiu) [inline platform/elrond/transaction.go:42]: `Memo` should be set to contain `tx.data` for regular transactions, correct?

File: platform/elrond/transaction_test.go [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+	"timestamp":1587715632,
+	"status":"Success"
+}`
+
+var rewardTxDst = blockatlas.Tx{
```
Comment (andreibancioiu) [inline platform/elrond/transaction_test.go]: Not `Dst`, but `Normalized` perhaps?

File: platform/elrond/transaction_test.go [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"testing"
+
+	"github.com/trustwallet/blockatlas/coin"
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+const rewardTxSrc = `
```
Comment (andreibancioiu) [inline platform/elrond/transaction_test.go]: Also add a test for regular transactions (not just reward transactions).

File: platform/elrond/transaction_test.go [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+		name:        "transfer",
+		apiResponse: rewardTxSrc,
+		expected:    &rewardTxDst,
+	})
+}
+
+func testNormalize(t *testing.T, _test *test) {
+	var tx Transaction
+	err := json.Unmarshal([]byte(_test.apiResponse), &tx)
+	if err != nil {
```
Comment (andreibancioiu) [inline platform/elrond/transaction_test.go:101]: You can use the usual `"github.com/stretchr/testify/assert"` in the tests. And `require.NotNil()` and so on.

**Review by andreibancioiu — COMMENTED** [review]
Make sure:
 - wait proxy PR to enter before opening the "real" PR to block atlas
 - make sure we run a final test (call BlockAtlas from a simple console application or curl) before opening the "real" PR

Then:
 - our fork / master should be up-to-date with the parent repository
 - create a new branch from the updated master, called `add-elrond` or so (the current one has a typo) and add your commit(s)
 - make sure everything is in order https://developer.trustwallet.com/blockatlas/newblockchain#pull-request
 - add a nice description to the PR, e.g. `Extended Block Atlas with blockchain & coin Elrond.` and state the two functions we've implemented for the Transactions API.

File: config.yml (hunk 2 above)
Comment (miiu96) [reply config.yml]: done.

File: platform/elrond/model.go (hunk 3 above)
Comment (miiu96) [reply platform/elrond/model.go]: done. now we have in transaction field Fee

File: platform/elrond/model.go (hunk 4 above)
Comment (miiu96) [reply platform/elrond/model.go]: renamed.

File: platform/elrond/transaction.go (hunk 5 above)
Comment (miiu96) [reply platform/elrond/transaction.go:49]: removed this part.

File: platform/registry.go (hunk 8 above)
Comment (miiu96) [reply platform/registry.go]: done.

File: platform/elrond/block.go (hunk 9 above)
Comment (miiu96) [reply platform/elrond/block.go]: renamed.

File: platform/elrond/client.go (hunk 10 above)
Comment (miiu96) [reply platform/elrond/client.go]: done.

File: platform/elrond/client.go (hunk 11 above)
Comment (miiu96) [reply platform/elrond/client.go]: done.

File: platform/elrond/client.go (hunk 12 above)
Comment (miiu96) [reply platform/elrond/client.go]: done.

File: platform/elrond/client.go (hunk 13 above)
Comment (miiu96) [reply platform/elrond/client.go]: done.

File: platform/elrond/model.go (hunk 14 above)
Comment (miiu96) [reply platform/elrond/model.go]: done.

File: platform/elrond/model.go (hunk 4 above)
Comment (miiu96) [reply platform/elrond/model.go]: done.

File: platform/elrond/client.go (hunk 15 above)
Comment (miiu96) [reply platform/elrond/client.go:49]: done.

File: platform/elrond/transaction.go (hunk 16 above)
Comment (miiu96) [reply platform/elrond/transaction.go:15]: done.

File: platform/elrond/transaction.go (hunk 17 above)
Comment (miiu96) [reply platform/elrond/transaction.go:42]: Yes you right. modified and set Memo with tx.Data

File: platform/elrond/transaction_test.go (hunk 18 above)
Comment (miiu96) [reply platform/elrond/transaction_test.go]: renamed.

File: platform/elrond/transaction_test.go (hunk 20 above)
Comment (miiu96) [reply platform/elrond/transaction_test.go:101]: done.

File: platform/elrond/transaction_test.go (hunk 19 above)
Comment (miiu96) [reply platform/elrond/transaction_test.go]: added.

File: README.md [hunk 21]
```diff
@@ -36,6 +36,7 @@ The observer API watches the chain for new transactions and generates notificati
 <a href="https://cosmos.network/" target="_blank"><img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/cosmos/info/logo.png" width="32" /></a>
 <a href="https://bitcoin.org/" target="_blank"><img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/bitcoin/info/logo.png" width="32" /></a>
 <a href="https://harmony.one/" target="_blank"><img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/harmony/info/logo.png" width="32" /></a>
+<a href="https://www.elrond.com/" target="_blank"><img src="https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/elrond/info/logo.png" width="32" /></a>
```
Comment (andreibancioiu) [inline README.md]: So, this should be the "official" URL of the logo. Will be added with a PR against the repository `trustwallet/assets`.

**Review by andreibancioiu — APPROVED** [review]

File: config.yml (hunk 1 above)
Comment (miiu96) [reply config.yml]: done.

File: platform/elrond/model.go (hunk 6 above)
Comment (miiu96) [reply platform/elrond/model.go]: done.

File: platform/elrond/transaction_test.go (hunk 7 above)
Comment (miiu96) [reply platform/elrond/transaction_test.go:21]: done.

File: platform/elrond/client.go [hunk 22]
```diff
@@ -29,20 +29,25 @@ func (c *Client) GetBlockByNumber(height int64) (*blockatlas.Block, error) {
...
 		ID:     block.Hash,
-		Txs:    NormalizeTxs(block.Transactions, ""),
+		Txs:    txs,
 	}, nil
 }
 
 func (c *Client) GetTxsOfAddress(address string) (blockatlas.TxPage, error) {
-	var bulkTxs BulkTransactions
+	var txPage TransactionsPage
+	// currently we fetch latest 20 transactions... in the future will may add pagination
```
Comment (AdoAdoAdo) [inline platform/elrond/client.go]: fix comment: 
will may add pagination -> will add pagination *or* may add pagination

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: platform/elrond/client.go (hunk 22 above)
Comment (miiu96) [reply platform/elrond/client.go]: comment changed

**Review by AdoAdoAdo — APPROVED** [review]

File: platform/elrond/model.go [hunk 23]
```diff
@@ -47,6 +59,14 @@ func (tx *Transaction) TxStatus() blockatlas.Status {
 	}
 }
 
+func (tx *Transaction) Fee() blockatlas.Amount {
```
Comment (andreibancioiu) [inline platform/elrond/model.go]: Let's not compute this here, in blockatlas. At least in proxy (if not in node).

Comment (miiu96) [reply platform/elrond/model.go]: done.

**Review by andreibancioiu — APPROVED** [review]

File: platform/elrond/model.go [hunk 24]
```diff
@@ -1,13 +1,24 @@
...
+type NetworkStatus struct {
+	NetworkStatus Status `json:"message"`
+}
+
+type Status struct {
+	Status StatusDetails `json:"status"`
+}
+
+type StatusDetails struct {
+	Round float64 `json:"erd_current_round"`
```
Comment (AdoAdoAdo) [inline platform/elrond/model.go:18]: snake case?

Comment (miiu96) [reply platform/elrond/model.go:18]: this is how route network/status/:shardID is implemented on proxy side

**Review by AdoAdoAdo — APPROVED** [review]
